src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.TryCloneFile.OSX.cs

```diff
@@ -0,0 +1,87 @@
+                // Success.
+                return true;
+            }
+            Interop.Error error = Interop.Sys.GetLastError();
+
+            // Some filesystems don't support ACLs, so may fail due to trying to copy ACLs.
+            // This will disable them and allow trying again (a maximum of 1 time).
+            if (error == Interop.Error.EINVAL && flags != 0)
+            {
+                flags = 0;
+                goto retry1;
+            }
+
+            // Try to delete the destination file if we're overwriting.
+            if (error == Interop.Error.EEXIST && overwrite)
+            {
+                // Delete the destination. This should fail on directories. Get a lock to the dest file to ensure we don't copy onto it when
+                // it's locked by something else, and then delete it. It should also fail if destination == source since it's already locked.
+                try
+                {
+                    using SafeFileHandle? dstHandle = SafeFileHandle.Open(destFullPath, FileMode.Open, FileAccess.ReadWrite,
+                        FileShare.None, FileOptions.None, preallocationSize: 0);
+                    if (Interop.Sys.Unlink(destFullPath) < 0)
+                    {
+                        Interop.Error errorInfo = Interop.Sys.GetLastError();
+                        if (error != Interop.Error.ENOENT)
+                        {
+                            // Fall back to standard copy as an unexpected error has occurred.
+                            return false;
+                        }
+                    }
+                }
+                catch (FileNotFoundException)
+                {
+                    // We don't want to throw if it's just the file not existing, since we're trying to delete it.
+                }
+
+                // Try clonefile now we've deleted the destination file.
+                retry2:
+                if (Interop.@libc.clonefile(sourceFullPath, destFullPath, flags) == 0)
+                {
+                    // Success.
+                    return true;
+                }
+                error = Interop.Sys.GetLastError();
+
+                // Some filesystems don't support ACLs, so may fail due to trying to copy ACLs.
+                // This will disable them and allow trying again (a maximum of 1 time).
+                if (error == Interop.Error.EINVAL && flags != 0)
+                {
+                    flags = 0;
+                    goto retry2;
+                }
+            }
+
+            // Check if it's not supported, if they're on different filesystems, the destination file still exists, or if we have unsupported flags still.
+            // EINVAL is checked again here since clonefile is implemented by filesystem-specific drivers, therefore it could potentially fail with flags=0.
+            if (error == Interop.Error.ENOTSUP || error == Interop.Error.EXDEV || error == Interop.Error.EEXIST || error == Interop.Error.EINVAL)
```

Comment:
> therefore it could potentially fail with flags=0

It musn't. The documentation says it's an OR of flags which makes zero is a valid value.

Can you leave out the `EINVAL` here? I want to avoid this code doing always taking the `EINVAL` path.

Reply:
Did you read my replies to this last time? It definitely could fail, because any filesystem can implement it however it wants, in which case we'd want to fall back. I don't think saving 1ns here for this comparison is worth someone possibly getting a random error saying "Invalid Parameter" and having no idea what it means.

Reply:
My point is that it could in theory fail, and if it did getting "Invalid Parameter" is not a great user experience. Adding 1 more value to compare here is not expensive compared to a copy operation anyway.

Reply:
> My point is that it could in theory fail (even though that would obviously be caused by an invalid filesystem implementation),

I wouldn't account for invalid filesystems which treat `0` an invalid value for flags.

I would account for us having a bug in the implementation that would be masked by `EINVAL` causing a fallback here.

Reply:
Should I add a debug statement instead then?